doc/synchronization.md

```diff
@@ -0,0 +1,356 @@
+and
+[2](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-2-of-2)
+- [JRuby's wiki page about concurrency](https://github.com/jruby/jruby/wiki/Concurrency-in-jruby)
+
+This memory model was created by: comparing
+[MRI](https://www.ruby-lang.org/en/), [JRuby](http://jruby.org/),
+[JRuby+Truffle](https://github.com/jruby/jruby/wiki/Truffle),
+[Rubinius](http://rubini.us/); taking account limitations of the implementations
+or their platforms; inspiration drawn from other existing memory models (Java,
+C++11). This is not a formal model.
+
+Key properties are:
+
+-   **volatility (V)** - Any written value is immediately visible to any
+    subsequent volatile reads including all writes leading to this value. (Same
+    meaning as in Java.)
+-   **atomicity (A)** - Operation is either done or not as a whole.
+-   **serialized (S)** - Operations are serialized in some order (they
+    cannot disappear). This is a new property not mentioned in other memory
+    models, since Java and C++ do not have dynamically defined fields. All
+    operations on one line in the table are serialized with each other.
+
+### Core behavior:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| local variable read/write/definition | - | x | x | Local variables are determined during parsing, they are not usually dinamically added (with exception of `local_variable_set`). Therefore definition is quite rare. |
+| instance variable read/write/(un)definition | - | x | x ||
+| class variable read/write/(un)definition | x | x | x ||
+| global variable read/write/definition | x | x | x | un-define us not possible currently. |
+| constant variable read/write/(un)definition | x | x | x ||
+| `Thread` local variable read/write/definition | - | x | x | un-define us not possible currently. |
+| `Fiber` local variable read/write/definition | - | x | x | un-define us not possible currently. |
+| method creation/redefinition/removal | x | x | x ||
+| include/extend | x | x | x | If `AClass` is included `AModule`, `AClass` gets all `AModule`'s methods at once. |
+
+
+Notes:
+
+-   Variable read reads value from preexisting variable.
+-   Variable definition creates new variable (operation is serialized with
+    writes, implies an update cannot be lost).
+-   Module/Class definition is actually constant definition. It is defined
+    instantly, however its methods are then processed sequentially.
+-   `||=`, `+=`, etc. are actually two operations read and write which implies
+    that it's not an atomic operation. See volatile variables
+    with compare-and-set.
+-   Method invocation does not have any special properties that includes
+    object initialization.
+
+Implementation differences from the model:
+
+-   MRI: everything is volatile.
+-   JRuby: `Thread` and `Fiber` local variables are volatile. Instance
+    variables are volatile on x86 and people may un/intentionally depend
+    on the fact.
```

Comment:
The fact that you can use `Thread#[]` or `Thread#thread_variable_get` on another thread seems like bad design. Indeed, not guaranteeing that it is volatile makes sense to me.